File: ~Programming_cheat_sheets/_compiled_languages/C/_data-structures-2020/stack/stack.c
#include <stdio.h>
#include <stdlib.h>

//
// Structs
//

typedef struct Node {
  int key;
  struct Node * next;
} Node;

typedef struct Stack {
  struct Node * top;
} Stack;

//
// functions
//

Node * createNode(int key){
  Node * node = (Node*)malloc(sizeof(Node));
  node->key = key;
  //node->next = NULL;
  return node;
}

void printNode(Node * node){
  if (node->next == NULL) {
    printf("Node{key=%i, next=NULL}\n", node->key);
    return;
  }
  printf("Node{key=%i, next=%i}\n", node->key, node->next->key);
}

Stack * createStack(){
  Stack * stack = (Stack*)malloc(sizeof(Stack));
  stack->top = NULL;
  return stack;
}

void add(Stack ** _stack, int key){
  Stack * stack = *_stack; // unwrap
  Node * node = createNode(key);
  if (stack->top == NULL) {
    stack->top = node;
    return;
  }

  node->next = stack->top;
  stack->top = node;
}

Node * rem(Stack ** _stack){
  Stack * stack = *_stack; // unwrap
  if (stack->top == NULL) {
    return NULL; // empty
  }

  Node * node = stack->top;
  stack->top = stack->top->next;
  return node;
}

void print(Stack * stack){
  Node * currentNode = stack->top;

  while (currentNode != NULL) {
    printNode(currentNode);
    currentNode = currentNode->next;
  }
}

/**
 * MAIN METHOD
 */
int main(){
  Stack * stack = createStack();
  add(&stack, 1);
  add(&stack, 2);
  add(&stack, 3);
  add(&stack, 4);
  add(&stack, 5);

  print(stack);

  printf("========== REMOVING:\n");

  printNode(rem(&stack));
  printNode(rem(&stack));
  printNode(rem(&stack));

  return 0;
}

/* OUTPUT:

Node{key=5, next=4}
Node{key=4, next=3}
Node{key=3, next=2}
Node{key=2, next=1}
Node{key=1, next=NULL}
========== REMOVING:
Node{key=5, next=4}
Node{key=4, next=3}
Node{key=3, next=2}

*/

File: ~other/~NEEDS_REVIEW/_scripting_languages/Python/data-structures-2020/stack/stack.py
# Stack

# ------------------------------ Node
class Node:
    def __init__(self, key):
        self.key = key
        self.next = None

# ------------------------------ Stack
class Stack:
    def __init__(self):
        self.head = None

    def push(self, key):
        node = Node(key)
        if self.head == None:
            self.head = node
            return

        node.next = self.head
        self.head = node

    def pop(self):
        if self.head == None:
            return

        self.head = self.head.next

    def show(self):
        currentNode = self.head
        while(currentNode != None):
            print(currentNode.key)
            currentNode = currentNode.next

# ------------------------------ Main
linkedList = Stack()
linkedList.push(1)
linkedList.push(2)
linkedList.push(3)
linkedList.push(4)
linkedList.push(5)
linkedList.push(6)


linkedList.pop()
linkedList.pop()
linkedList.pop()

linkedList.show()
